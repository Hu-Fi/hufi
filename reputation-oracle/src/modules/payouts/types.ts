import { IEscrow } from '@human-protocol/sdk';

export type CampaignManifest = {
  type: string;
  daily_volume_target: number;
  exchange: string;
  pair: string;
  fund_token: string;
  start_date: string;
  end_date: string;
};

export type CampaignWithResults = Required<
  Pick<
    IEscrow,
    | 'chainId'
    | 'address'
    | 'manifestUrl'
    | 'manifestHash'
    | 'intermediateResultsUrl'
  > & {
    fundTokenAddress: string;
    fundTokenDecimals: number;
    fundAmount: number;
  }
>;

export type ParticipantOutcome = {
  address: string;
  score: number;
  total_volume: number;
};

export type ParticipantsOutcomesBatch = {
  id: string;
  results: ParticipantOutcome[];
};

export type IntermediateResult = {
  from: Date;
  to: Date;
  total_volume: number;
  participants_outcomes_batches: ParticipantsOutcomesBatch[];
};

export type IntermediateResultsData = {
  chain_id: number;
  address: string;
  exchange: string;
  pair: string;
  results: IntermediateResult[];
};

export type CalculatedReward = {
  address: string;
  amount: number;
};

export type CalculatedRewardsBatch = {
  id: string;
  rewards: CalculatedReward[];
};
