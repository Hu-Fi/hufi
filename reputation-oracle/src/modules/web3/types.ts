import type { Provider, Wallet } from 'ethers';

import type { ChainId } from '@/common/constants';

export type Chain = {
  id: ChainId;
  rpcUrl: string;
};

export type WalletWithProvider = Wallet & { provider: Provider };
