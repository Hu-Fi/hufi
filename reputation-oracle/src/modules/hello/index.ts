export { HelloService } from './hello.service';
export { HelloModule } from './hello.module';
