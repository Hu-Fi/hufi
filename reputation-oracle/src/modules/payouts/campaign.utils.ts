import Joi from 'joi';

import { ChainIds } from '@/common/constants';
import * as httpUtils from '@/common/utils/http';

import { IntermediateResults } from './types';

const participantOutcome = Joi.object({
  address: Joi.string().required(),
  score: Joi.number().required(),
  total_volume: Joi.number().min(0).required(),
});

const participantsOutcomesBatchSchema = Joi.object({
  id: Joi.string().uuid({ version: 'uuidv4' }).required(),
  results: Joi.array().items(participantOutcome).required(),
});

const intermediateResultSchema = Joi.object({
  from: Joi.date().iso().required(),
  to: Joi.date().iso().greater(Joi.ref('from')).required(),
  total_volume: Joi.number().positive().required(),
  participants_outcomes_batches: Joi.array()
    .items(participantsOutcomesBatchSchema)
    .required(),
});

const intermedateResultsSchema = Joi.object({
  chain_id: Joi.number().valid(ChainIds).required(),
  address: Joi.string().required(),
  exchange: Joi.string().required(),
  pair: Joi.string()
    .pattern(/^[A-Z]{3,10}\/[A-Z]{3,10}$/)
    .required(),
  results: Joi.array().items(intermediateResultSchema).required(),
}).options({ allowUnknown: true, stripUnknown: true });

export async function downloadIntermediateResults(
  url: string,
): Promise<IntermediateResults> {
  const resultsData = await httpUtils.downloadFile(url);

  try {
    const resultsJson = JSON.parse(resultsData.toString());
    const valudatedResults = Joi.attempt(resultsJson, intermedateResultsSchema);

    return valudatedResults;
  } catch {
    throw new Error('Invalid intermediate results schema');
  }
}
