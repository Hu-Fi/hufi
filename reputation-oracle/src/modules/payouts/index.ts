export { PayoutsService } from './payouts.service';
export { PayoutModule } from './payouts.module';
