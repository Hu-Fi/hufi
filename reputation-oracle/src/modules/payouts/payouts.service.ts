import crypto from 'crypto';

import {
  EscrowClient,
  EscrowStatus,
  EscrowUtils,
  IEscrowsFilter,
} from '@human-protocol/sdk';
import { Injectable } from '@nestjs/common';
import { ethers } from 'ethers';

import type { ChainId } from '@/common/constants';
import { ContentType } from '@/common/enums';
import * as decimalUtils from '@/common/utils/decimal';
import * as web3Utils from '@/common/utils/web3';
import { Web3ConfigService } from '@/config';
import logger from '@/logger';
import { Web3Service } from '@/modules/web3';

import * as campaignUtils from './campaign.utils';
import {
  CalculatedRewardsBatch,
  CampaignWithResults,
  IntermediateResult,
  IntermediateResultsData,
} from './types';
import { StorageService } from '../storage';

@Injectable()
export class PayoutsService {
  private readonly logger = logger.child({
    context: PayoutsService.name,
  });

  constructor(
    private readonly storageService: StorageService,
    private readonly web3ConfigService: Web3ConfigService,
    private readonly web3Service: Web3Service,
  ) {}

  async runPayoutsCycle(): Promise<void> {
    this.logger.info('Going to run payouts cycle', {
      supportedChainIds: this.web3Service.supportedChainIds,
    });

    for (const chainId of this.web3Service.supportedChainIds) {
      const campaigns = await this.getCampaignsForPayouts(chainId);

      logger.info('Found campaigns waiting for payouts', {
        chainId,
        campaigns: campaigns.map((c) => c.address),
      });

      for (const campaign of campaigns) {
        await this.runPayoutsCycleForCampaign(campaign);
      }
    }

    this.logger.info('Payouts cycle finished');
  }

  async runPayoutsCycleForCampaign(
    campaign: CampaignWithResults,
  ): Promise<void> {
    const logger = this.logger.child({
      chainId: campaign.chainId,
      campaign: campaign.address,
    });

    logger.info('Run payouts cycle for campaign');

    try {
      const signer = this.web3Service.getSigner(campaign.chainId);
      const escrowClient = await EscrowClient.build(signer);

      let allResultsPaid = false;
      try {
        const [manifest, intermediateResultsData] = await Promise.all([
          campaignUtils.downloadCampaignManifest(
            campaign.manifestUrl,
            campaign.manifestHash,
          ),
          campaignUtils.downloadIntermediateResults(
            campaign.intermediateResultsUrl,
          ),
        ]);

        if (intermediateResultsData.results.length === 0) {
          throw new Error('Intermediate results are not recorded');
        }

        const dailyReward = campaignUtils.calculateDailyReward(
          campaign,
          manifest,
        );
        logger.info('Calculated daily reward for campaign', {
          startDate: manifest.start_date,
          endDate: manifest.end_date,
          fundAmount: campaign.fundAmount,
          fundTokenAddress: campaign.fundTokenAddress,
          dailyReward,
        });

        const finalResultsMeta = await this.uploadFinalResults(
          campaign,
          intermediateResultsData,
        );

        for (const intermediateResult of intermediateResultsData.results) {
          /**
           * TODO: somehow check if this result was already paid before
           */
          if (intermediateResult.total_volume === 0) {
            continue;
          }

          const rewardsBatches = this.calculateRewardsForDailyResult(
            intermediateResult,
            dailyReward,
            manifest.daily_volume_target,
          );

          logger.info('Rewards calculated', {
            rewardsBatches,
          });

          for (const rewardsBatch of rewardsBatches) {
            /**
             * TODO: somehow check if this rewards batch was already paid before
             */

            const recipientToAmountMap = new Map<string, bigint>();
            for (const { address, amount } of rewardsBatch.rewards) {
              recipientToAmountMap.set(
                address,
                ethers.parseUnits(
                  amount.toString(),
                  campaign.fundTokenDecimals,
                ),
              );
            }

            const gasPrice = await this.web3Service.calculateGasPrice(
              campaign.chainId,
            );

            await escrowClient.bulkPayOut(
              campaign.address,
              Array.from(recipientToAmountMap.keys()),
              Array.from(recipientToAmountMap.values()),
              finalResultsMeta.url,
              finalResultsMeta.hash,
              /**
               * TODO: replace it with batch id when SDK is updated
               */
              1,
              false,
              {
                gasPrice,
              },
            );
          }
        }

        const lastResultsAt = intermediateResultsData.results
          .at(-1)
          ?.to.toISOString();

        if (lastResultsAt === manifest.end_date) {
          allResultsPaid = true;
        }

        logger.info('Finished payouts for campaign');
      } catch (error) {
        logger.error('Payouts failed for campaign', error);
      }

      if (allResultsPaid) {
        try {
          await escrowClient.complete(campaign.address);
        } catch (error) {
          logger.error('Failed to complete campaign', error);
        }
      } else {
        logger.info('Campaign not finished yet, skip completion');
      }

      logger.info('Finished payouts cycle for campaign');
    } catch (error) {
      logger.error('Error while running payouts cycle for campaign', error);
    }
  }

  private calculateRewardsForDailyResult(
    intermediateResult: IntermediateResult,
    dailyReward: number,
    dailyVolumeTarget: number,
  ): CalculatedRewardsBatch[] {
    const rewardRatio = Math.min(
      intermediateResult.total_volume / dailyVolumeTarget,
      1,
    );

    const totalSharedReward = rewardRatio * dailyReward;

    let totalScore = 0;
    for (const outcomesBatch of intermediateResult.participants_outcomes_batches) {
      for (const outcome of outcomesBatch.results) {
        totalScore += outcome.score;
      }
    }

    const rewardsBatches: CalculatedRewardsBatch[] = [];
    for (const outcomesBatch of intermediateResult.participants_outcomes_batches) {
      const rewardsBatch: CalculatedRewardsBatch = {
        id: outcomesBatch.id,
        rewards: [],
      };

      for (const outcome of outcomesBatch.results) {
        const participantShare = decimalUtils.div(totalScore, outcome.score);
        const rewardAmount = decimalUtils.div(
          totalSharedReward,
          participantShare,
        );

        rewardsBatch.rewards.push({
          address: outcome.address,
          amount: rewardAmount,
        });
      }

      rewardsBatches.push(rewardsBatch);
    }

    return rewardsBatches;
  }

  private async uploadFinalResults(
    campaign: CampaignWithResults,
    finalResults: IntermediateResultsData,
  ): Promise<{
    url: string;
    hash: string;
  }> {
    const stringifiedResults = JSON.stringify(finalResults);
    const resultsHash = crypto
      .createHash('sha256')
      .update(stringifiedResults)
      .digest('hex');

    const fileName = `${campaign.address}.json`;

    const resultsUrl = await this.storageService.uploadData(
      stringifiedResults,
      fileName,
      ContentType.JSON,
    );

    return {
      url: resultsUrl,
      hash: resultsHash,
    };
  }

  private async getCampaignsForPayouts(
    chainId: ChainId,
  ): Promise<CampaignWithResults[]> {
    const baseFilter: IEscrowsFilter = {
      chainId: chainId as number,
      reputationOracle: this.web3ConfigService.operatorAddress,
    };

    /**
     * TODO: add "toCancelEscrows" when escrow cancelletion is done
     */
    const [pendingEscrows, partialEscrows] = await Promise.all([
      EscrowUtils.getEscrows({
        ...baseFilter,
        status: EscrowStatus.Pending,
      }),
      EscrowUtils.getEscrows({
        ...baseFilter,
        status: EscrowStatus.Partial,
      }),
    ]);

    const campaignsWithResults: CampaignWithResults[] = [];
    for (const escrow of [...pendingEscrows, ...partialEscrows]) {
      if (!escrow.intermediateResultsUrl) {
        continue;
      }

      const fundTokenDecimals = web3Utils.getTokenDecimals(
        escrow.chainId,
        escrow.token,
      );

      campaignsWithResults.push({
        chainId: escrow.chainId,
        address: escrow.address,
        manifestUrl: escrow.manifestUrl as string,
        manifestHash: escrow.manifestHash as string,
        intermediateResultsUrl: escrow.intermediateResultsUrl as string,
        fundTokenAddress: escrow.token,
        fundTokenDecimals,
        fundAmount: Number(
          ethers.formatUnits(escrow.totalFundedAmount, fundTokenDecimals),
        ),
      });
    }

    return campaignsWithResults;
  }
}
