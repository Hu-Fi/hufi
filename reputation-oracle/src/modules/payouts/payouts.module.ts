import { Module } from '@nestjs/common';

import { StorageModule } from '@/modules/storage';
import { Web3Module } from '@/modules/web3';

import { PayoutsService } from './payouts.service';

@Module({
  imports: [StorageModule, Web3Module],
  providers: [PayoutsService],
  exports: [PayoutsService],
})
export class PayoutModule {}
