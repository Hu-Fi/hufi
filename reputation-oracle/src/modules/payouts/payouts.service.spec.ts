/* eslint-disable @typescript-eslint/no-explicit-any */
jest.mock('@human-protocol/sdk');
jest.mock('@/logger');

import crypto from 'crypto';

import { faker } from '@faker-js/faker';
import { createMock } from '@golevelup/ts-jest';
import { EscrowClient, EscrowStatus, EscrowUtils } from '@human-protocol/sdk';
import { Test } from '@nestjs/testing';
import Decimal from 'decimal.js';
import { ethers } from 'ethers';
import _ from 'lodash';

import { ContentType } from '@/common/enums';
import { Web3ConfigService } from '@/config';
import logger from '@/logger';
import { StorageService } from '@/modules/storage';
import { Web3Service } from '@/modules/web3';
import {
  generateTestnetChainId,
  mockWeb3ConfigService,
} from '@/modules/web3/fixtures';

import {
  generateCampaign,
  generateEscrow,
  generateIntermediateResult,
  generateIntermediateResultsData,
  generateManifest,
  generateParticipantOutcome,
} from './fixtures';
import precisionSensitiveResult from './fixtures/precision_sensitive_result.json';
import { PayoutsService } from './payouts.service';
import * as payoutsUtils from './payouts.utils';
import { CampaignWithResults, IntermediateResult } from './types';

const mockStorageService = createMock<StorageService>();
const mockWeb3Service = createMock<Web3Service>();

const mockedEscrowClient = jest.mocked(EscrowClient);
const mockedEscrowUtils = jest.mocked(EscrowUtils);

describe('PayoutsService', () => {
  let payoutsService: PayoutsService;

  beforeAll(async () => {
    const moduleRef = await Test.createTestingModule({
      providers: [
        PayoutsService,
        {
          provide: StorageService,
          useValue: mockStorageService,
        },
        {
          provide: Web3Service,
          useValue: mockWeb3Service,
        },
        {
          provide: Web3ConfigService,
          useValue: mockWeb3ConfigService,
        },
      ],
    }).compile();

    payoutsService = moduleRef.get<PayoutsService>(PayoutsService);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should be defined', () => {
    expect(payoutsService).toBeDefined();
  });

  describe('getCampaignsForPayouts', () => {
    const TEST_TOKEN_DECIMALS = faker.helpers.arrayElement([6, 18]);

    let chainId: number;

    beforeEach(() => {
      chainId = generateTestnetChainId();
      mockWeb3Service.getTokenDecimals.mockResolvedValue(TEST_TOKEN_DECIMALS);
    });

    it('should query escrows with correct params', async () => {
      mockedEscrowUtils.getEscrows.mockResolvedValueOnce([]);

      await payoutsService['getCampaignsForPayouts'](chainId);

      expect(mockedEscrowUtils.getEscrows).toHaveBeenCalledTimes(1);
      expect(mockedEscrowUtils.getEscrows).toHaveBeenCalledWith({
        chainId,
        reputationOracle: mockWeb3ConfigService.operatorAddress,
        status: [EscrowStatus.Pending, EscrowStatus.Partial],
        first: 100,
      });
    });

    it('shold return campaign data from escrow', async () => {
      const expectedEscrow = generateEscrow();
      mockedEscrowUtils.getEscrows.mockResolvedValueOnce([expectedEscrow]);

      const campaigns = await payoutsService['getCampaignsForPayouts'](chainId);

      expect(campaigns.length).toBe(1);
      expect(campaigns[0]).toEqual({
        chainId: expectedEscrow.chainId,
        address: expectedEscrow.address,
        manifest: expectedEscrow.manifest,
        manifestHash: expectedEscrow.manifestHash,
        intermediateResultsUrl: expectedEscrow.intermediateResultsUrl,
        fundTokenAddress: expectedEscrow.token,
        fundTokenDecimals: TEST_TOKEN_DECIMALS,
        fundAmount: Number(
          ethers.formatUnits(
            expectedEscrow.totalFundedAmount,
            TEST_TOKEN_DECIMALS,
          ),
        ),
        launcher: expectedEscrow.launcher,
      });
    });

    it('should skip escrows w/o intermediate results', async () => {
      const expectedEscrow = generateEscrow();

      mockedEscrowUtils.getEscrows.mockResolvedValueOnce([
        expectedEscrow,
        Object.assign(generateEscrow(), { intermediateResultsUrl: '' }),
      ]);

      const campaigns = await payoutsService['getCampaignsForPayouts'](chainId);

      expect(campaigns.length).toBe(1);
      expect(campaigns[0].address).toEqual(expectedEscrow.address);
    });
  });

  describe('uploadFinalResults', () => {
    it('should upload results with proper meta', async () => {
      const expectedFileUrl = faker.internet.url();
      mockStorageService.uploadData.mockResolvedValueOnce(expectedFileUrl);

      const intermediateResultsData = generateIntermediateResultsData();
      const stringifiedResults = JSON.stringify(intermediateResultsData);
      const resultsHash = crypto
        .createHash('sha256')
        .update(stringifiedResults)
        .digest('hex');

      const campaignAddress = ethers.getAddress(
        faker.finance.ethereumAddress(),
      );

      const meta = await payoutsService['uploadFinalResults'](
        {
          address: campaignAddress,
        } as CampaignWithResults,
        intermediateResultsData,
      );

      expect(mockStorageService.uploadData).toHaveBeenCalledTimes(1);
      expect(mockStorageService.uploadData).toHaveBeenCalledWith(
        stringifiedResults,
        `${campaignAddress}/${resultsHash}.json`,
        ContentType.JSON,
      );

      expect(meta).toEqual({
        url: expectedFileUrl,
        hash: resultsHash,
      });
    });
  });

  describe('runPayoutsCycle', () => {
    const originalSupportedChainIds = mockWeb3Service.supportedChainIds;
    const supportedChainId = faker.number.int();

    let spyOnGetCampaignsForPayouts: jest.SpyInstance;
    let spyOnRunPayoutsCycleForCampaign: jest.SpyInstance;

    beforeAll(() => {
      spyOnGetCampaignsForPayouts = jest.spyOn(
        payoutsService as any,
        'getCampaignsForPayouts',
      );
      spyOnGetCampaignsForPayouts.mockImplementation();

      spyOnRunPayoutsCycleForCampaign = jest.spyOn(
        payoutsService as any,
        'runPayoutsCycleForCampaign',
      );
      spyOnRunPayoutsCycleForCampaign.mockImplementation();
    });

    afterAll(() => {
      (mockWeb3Service.supportedChainIds as any) = originalSupportedChainIds;

      spyOnGetCampaignsForPayouts.mockRestore();
      spyOnRunPayoutsCycleForCampaign.mockRestore();
    });

    beforeEach(() => {
      (mockWeb3Service.supportedChainIds as any) = [supportedChainId];
    });

    it('should process each supported chain', async () => {
      (mockWeb3Service.supportedChainIds as any) = [
        faker.number.int(),
        faker.number.int(),
      ];
      spyOnGetCampaignsForPayouts.mockResolvedValue([]);

      await payoutsService.runPayoutsCycle();

      expect(spyOnGetCampaignsForPayouts).toHaveBeenCalledTimes(2);
      for (const chainId of mockWeb3Service.supportedChainIds) {
        expect(spyOnGetCampaignsForPayouts).toHaveBeenCalledWith(chainId);
      }

      expect(spyOnRunPayoutsCycleForCampaign).toHaveBeenCalledTimes(0);
    });

    it('should process each campaign for chain', async () => {
      const campaigns = [generateCampaign(), generateCampaign()];
      spyOnGetCampaignsForPayouts.mockResolvedValueOnce(campaigns);

      await payoutsService.runPayoutsCycle();

      expect(spyOnGetCampaignsForPayouts).toHaveBeenCalledTimes(1);

      expect(spyOnRunPayoutsCycleForCampaign).toHaveBeenCalledTimes(
        campaigns.length,
      );
      for (const campaign of campaigns) {
        expect(spyOnRunPayoutsCycleForCampaign).toHaveBeenCalledWith(campaign);
      }
    });
  });

  describe('calculateRewardsForIntermediateResult', () => {
    const TEST_TOKEN_DECIMALS = faker.number.int({ min: 5, max: 18 });

    it('should return rewards in batches', () => {
      const rewardsBatches = payoutsService[
        'calculateRewardsForIntermediateResult'
      ](generateIntermediateResult(), TEST_TOKEN_DECIMALS);

      expect(rewardsBatches.length).toBe(0);
    });

    it('should calculate rewards for all participants in batches', () => {
      const intermediateResult = generateIntermediateResult();

      const nBatches = faker.number.int({ min: 2, max: 4 });
      let totalParticipants = 0;
      const participantAddressesSet = new Set<string>();
      for (let i = 0; i < nBatches; i += 1) {
        const batch = {
          id: faker.string.uuid(),
          results: Array.from(
            { length: faker.number.int({ min: 1, max: 3 }) },
            () => {
              const outcome = generateParticipantOutcome({ score: 1 });

              participantAddressesSet.add(outcome.address);

              return outcome;
            },
          ),
        };

        intermediateResult.participants_outcomes_batches.push(batch);

        totalParticipants += batch.results.length;
      }

      const equalReward = faker.number.int({ min: 15, max: 42 });
      intermediateResult.reserved_funds = totalParticipants * equalReward;

      const rewardsBatches = payoutsService[
        'calculateRewardsForIntermediateResult'
      ](intermediateResult, TEST_TOKEN_DECIMALS);

      const rewardsMap = new Map<string, number>();
      for (
        let i = 0;
        i < intermediateResult.participants_outcomes_batches.length;
        i += 1
      ) {
        const { id: rewardsBatchId, rewards: rewardsBatch } = rewardsBatches[i];

        expect(rewardsBatchId).toBe(
          intermediateResult.participants_outcomes_batches[i].id,
        );

        for (const reward of rewardsBatch) {
          rewardsMap.set(reward.address, reward.amount);
        }
      }

      for (const participant of participantAddressesSet) {
        expect(rewardsMap.get(participant)).toBe(equalReward);
      }
    });

    it('should correctly calculate rewards for precision-sensitive results', () => {
      const rewards = payoutsService['calculateRewardsForIntermediateResult'](
        Object.assign({}, precisionSensitiveResult, {
          from: new Date(precisionSensitiveResult.from),
          to: new Date(precisionSensitiveResult.to),
        }),
        18,
      );

      let total = new Decimal(0);
      for (const reward of rewards[0].rewards) {
        total = total.plus(reward.amount);
      }
      expect(total.toNumber()).toBe(41.99999999999999);
      /**
       * Input and shanpshot for this test are based on real-data
       * and correct at the time of adding. They shouldn't be changed
       * unless some bug in input/output itself is found.
       */
      expect(rewards).toMatchSnapshot();
    });
  });

  describe('runPayoutsCycleForCampaign', () => {
    const mockedCampaign = generateCampaign();
    const mockedGasPrice = faker.number.bigInt();
    const mockedParticipantAddress = faker.finance.ethereumAddress();
    const mockedParticipantsOutcomesBatch = {
      id: faker.string.uuid(),
      results: [
        {
          address: mockedParticipantAddress,
          score: faker.number.float({ min: 0.1 }),
        },
      ],
    };
    const mockedReservedFunds = faker.number.int({ min: 1, max: 10 });
    const mockedFinalResultsUrl = faker.internet.url();
    const mockedFinalResultsHash = faker.string.hexadecimal();
    const mockedManifest = generateManifest();

    let spyOnRetrieveCampaignManifest: jest.SpyInstance;
    let spyOnDownloadIntermediateResults: jest.SpyInstance;
    let spyOnUploadFinalResults: jest.SpyInstance;
    let spyOnGetBulkPayoutsCount: jest.SpyInstance;
    let spyOnWriteRewardsBatchToFile: jest.SpyInstance;

    const mockedGetEscrowBalance = jest.fn();
    const mockedGetEscrowReservedFunds = jest.fn();
    const mockedGetEscrowStatus = jest.fn();
    const mockedBulkPayOut = jest.fn();
    const mockedCompleteEscrow = jest.fn();

    let mockedIntermediateResult: IntermediateResult;
    let mockedEscrowReservedFunds: bigint;

    beforeAll(() => {
      spyOnRetrieveCampaignManifest = jest.spyOn(
        payoutsUtils,
        'retrieveCampaignManifest',
      );
      spyOnRetrieveCampaignManifest.mockImplementation();

      spyOnDownloadIntermediateResults = jest.spyOn(
        payoutsUtils,
        'downloadIntermediateResults',
      );
      spyOnDownloadIntermediateResults.mockImplementation();

      spyOnUploadFinalResults = jest.spyOn(
        payoutsService as any,
        'uploadFinalResults',
      );
      spyOnUploadFinalResults.mockImplementation();

      spyOnGetBulkPayoutsCount = jest.spyOn(
        payoutsService as any,
        'getBulkPayoutsCount',
      );
      spyOnGetBulkPayoutsCount.mockImplementation();

      spyOnWriteRewardsBatchToFile = jest.spyOn(
        payoutsService as any,
        'writeRewardsBatchToFile',
      );
      spyOnWriteRewardsBatchToFile.mockImplementation();
    });

    afterAll(() => {
      spyOnRetrieveCampaignManifest.mockRestore();
      spyOnDownloadIntermediateResults.mockRestore();
      spyOnUploadFinalResults.mockRestore();
      spyOnGetBulkPayoutsCount.mockRestore();
      spyOnWriteRewardsBatchToFile.mockRestore();
    });

    beforeEach(() => {
      mockedEscrowClient.build.mockResolvedValue({
        getBalance: mockedGetEscrowBalance,
        getStatus: mockedGetEscrowStatus,
        getReservedFunds: mockedGetEscrowReservedFunds,
        bulkPayOut: mockedBulkPayOut,
        complete: mockedCompleteEscrow,
      } as unknown as EscrowClient);

      mockWeb3Service.calculateGasPrice.mockResolvedValue(mockedGasPrice);

      spyOnRetrieveCampaignManifest.mockResolvedValueOnce(mockedManifest);

      mockedIntermediateResult = generateIntermediateResult();
      mockedIntermediateResult.reserved_funds = mockedReservedFunds;
      mockedIntermediateResult.participants_outcomes_batches.push(
        _.cloneDeep(mockedParticipantsOutcomesBatch),
      );
      const intermediateResultsData = generateIntermediateResultsData();
      intermediateResultsData.results.push(mockedIntermediateResult);
      spyOnDownloadIntermediateResults.mockResolvedValueOnce(
        intermediateResultsData,
      );

      mockedEscrowReservedFunds = ethers.parseUnits(
        mockedReservedFunds.toString(),
        mockedCampaign.fundTokenDecimals,
      );
      mockedGetEscrowReservedFunds.mockResolvedValue(mockedEscrowReservedFunds);

      spyOnGetBulkPayoutsCount.mockResolvedValueOnce(0);

      spyOnUploadFinalResults.mockResolvedValueOnce({
        url: mockedFinalResultsUrl,
        hash: mockedFinalResultsHash,
      });
    });

    it('should gracefully handle incomplete intermediate results', async () => {
      spyOnDownloadIntermediateResults
        .mockReset()
        .mockResolvedValueOnce(generateIntermediateResultsData());

      await payoutsService.runPayoutsCycleForCampaign(mockedCampaign);

      expect(logger.error).toHaveBeenCalledTimes(1);
      expect(logger.error).toHaveBeenCalledWith(
        'Payouts failed for campaign',
        new Error('Intermediate results are not recorded'),
      );

      expect(spyOnGetBulkPayoutsCount).toHaveBeenCalledTimes(0);
      expect(spyOnUploadFinalResults).toHaveBeenCalledTimes(0);
      expect(mockedBulkPayOut).toHaveBeenCalledTimes(0);
      expect(mockedCompleteEscrow).toHaveBeenCalledTimes(0);
    });

    it('should gracefully handle invalid reserved funds', async () => {
      mockedGetEscrowReservedFunds.mockReset().mockResolvedValueOnce(
        faker.number.int({
          max:
            mockedIntermediateResult.reserved_funds -
            faker.number.float({ min: 0.0000001 }),
        }),
      );

      await payoutsService.runPayoutsCycleForCampaign(mockedCampaign);

      expect(logger.error).toHaveBeenCalledTimes(1);
      expect(logger.error).toHaveBeenCalledWith(
        'Payouts failed for campaign',
        new Error('Expected payouts amount higher than reserved funds'),
      );
      expect(mockedBulkPayOut).toHaveBeenCalledTimes(0);
      expect(mockedCompleteEscrow).toHaveBeenCalledTimes(0);
    });

    it('should run payouts and not attempt finish campaign if not ended', async () => {
      await payoutsService.runPayoutsCycleForCampaign(mockedCampaign);

      expect(mockedBulkPayOut).toHaveBeenCalledTimes(1);
      expect(mockedBulkPayOut).toHaveBeenCalledWith(
        mockedCampaign.address,
        [mockedParticipantAddress],
        [
          ethers.parseUnits(
            mockedReservedFunds.toString(),
            mockedCampaign.fundTokenDecimals,
          ),
        ],
        mockedFinalResultsUrl,
        mockedFinalResultsHash,
        mockedParticipantsOutcomesBatch.id,
        false,
        {
          gasPrice: mockedGasPrice,
        },
      );

      expect(logger.info).toHaveBeenCalledTimes(6);
      expect(logger.info).toHaveBeenCalledWith(
        'Rewards batch successfully paid',
        {
          batchId: mockedIntermediateResult.participants_outcomes_batches[0].id,
        },
      );
      expect(logger.info).toHaveBeenCalledWith(
        'Campaign not finished yet, skip completion',
      );

      expect(mockedGetEscrowReservedFunds).toHaveBeenCalledTimes(1);
      expect(mockedGetEscrowReservedFunds).toHaveBeenCalledWith(
        mockedCampaign.address,
      );

      expect(mockedGetEscrowStatus).toHaveBeenCalledTimes(0);
      expect(mockedCompleteEscrow).toHaveBeenCalledTimes(0);
    });

    it('should skip already paid results', async () => {
      spyOnGetBulkPayoutsCount.mockReset().mockResolvedValueOnce(1);

      await payoutsService.runPayoutsCycleForCampaign(mockedCampaign);

      expect(logger.debug).toHaveBeenCalledTimes(2);
      expect(logger.debug).toHaveBeenCalledWith(
        'Skipped rewards batch as per bulkPayoutsCount',
        {
          batchId: mockedIntermediateResult.participants_outcomes_batches[0].id,
          batchTotalReward: expect.stringMatching(/^\d+(\.\d+)?$/),
        },
      );

      expect(mockedBulkPayOut).toHaveBeenCalledTimes(0);
      expect(mockedCompleteEscrow).toHaveBeenCalledTimes(0);
    });

    describe('campaign completion', () => {
      beforeEach(() => {
        const manifest = generateManifest();
        manifest.end_date = mockedIntermediateResult.to.toISOString();
        spyOnRetrieveCampaignManifest
          .mockReset()
          .mockResolvedValueOnce(manifest);
      });

      it('should run payouts and skip completion if auto-completed', async () => {
        mockedGetEscrowStatus.mockResolvedValueOnce(EscrowStatus.Complete);

        await payoutsService.runPayoutsCycleForCampaign(mockedCampaign);

        expect(logger.info).toHaveBeenCalledTimes(6);
        expect(logger.info).toHaveBeenCalledWith(
          'Campaign auto-completed during payouts',
        );

        expect(mockedBulkPayOut).toHaveBeenCalledTimes(1);
        expect(mockedCompleteEscrow).toHaveBeenCalledTimes(0);
      });

      it('should run payouts and complete campaign if not auto-completed', async () => {
        mockedGetEscrowStatus.mockResolvedValueOnce(EscrowStatus.Partial);

        await payoutsService.runPayoutsCycleForCampaign(mockedCampaign);

        expect(logger.info).toHaveBeenCalledTimes(6);
        expect(logger.info).toHaveBeenCalledWith(
          'Campaign is fully paid, completing it',
        );

        expect(mockedBulkPayOut).toHaveBeenCalledTimes(1);
        expect(mockedCompleteEscrow).toHaveBeenCalledTimes(1);
      });

      it('should warn if unknown status', async () => {
        const unknownStatus = -1;
        mockedGetEscrowStatus.mockResolvedValueOnce(unknownStatus);

        await payoutsService.runPayoutsCycleForCampaign(mockedCampaign);

        expect(logger.warn).toHaveBeenCalledTimes(1);
        expect(logger.warn).toHaveBeenCalledWith(
          'Unexpected campaign escrow status',
          {
            escrowStatus: unknownStatus,
          },
        );

        expect(mockedBulkPayOut).toHaveBeenCalledTimes(1);
        expect(mockedCompleteEscrow).toHaveBeenCalledTimes(0);
      });

      it('should complete with refund if campaign has "zero" results', async () => {
        const mockedEscrowBalance = faker.number.bigInt({ min: 1 });
        mockedGetEscrowReservedFunds.mockReset().mockResolvedValueOnce(0);
        mockedIntermediateResult.reserved_funds = 0;
        mockedIntermediateResult.participants_outcomes_batches = [];

        mockedGetEscrowStatus.mockResolvedValueOnce(EscrowStatus.Pending);
        mockedGetEscrowBalance.mockResolvedValueOnce(mockedEscrowBalance);

        await payoutsService.runPayoutsCycleForCampaign(mockedCampaign);

        expect(logger.info).toHaveBeenCalledTimes(5);
        expect(logger.info).toHaveBeenCalledWith(
          'Campaign ended with empty results, completing it',
        );

        expect(mockedBulkPayOut).toHaveBeenCalledTimes(1);
        expect(mockedBulkPayOut).toHaveBeenLastCalledWith(
          mockedCampaign.address,
          [mockedCampaign.launcher],
          [mockedEscrowBalance],
          mockedFinalResultsUrl,
          mockedFinalResultsHash,
          'empty_results_tx',
          true,
          {
            gasPrice: mockedGasPrice,
          },
        );
      });
    });
  });
});
