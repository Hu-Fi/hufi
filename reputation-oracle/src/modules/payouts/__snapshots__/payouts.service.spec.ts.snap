// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PayoutsService calculateRewardsForIntermediateResult should correctly calculate rewards for precision-sensitive results 1`] = `
[
  {
    "id": "db634797-f8c2-44ce-b8ac-4c09038a7bdd",
    "rewards": [
      {
        "address": "0x0000000000000000000000000000000000000001",
        "amount": 0.9756486072118352,
      },
      {
        "address": "0x0000000000000000000000000000000000000002",
        "amount": 41.02435139278816,
      },
    ],
  },
]
`;
